lib/apiservers/engine/backends/container.go

```diff
@@ -359,123 +451,167 @@ func (c *ContainerBackend) ContainerExecStart(ctx context.Context, eid string, s
 	id := vc.ContainerID
 	name := vc.Name
 
-	op.Debugf("exec start of %s in container %d", eid, id)
+	op.Debugf("Exec start of %s in container %s", eid, id)
 
+	var ec *models.TaskInspectResponse
 	operation := func() error {
-		handle, err := c.Handle(id, name)
-		if err != nil {
-			op.Errorf("Failed to obtain handle during exec start for container(%s) due to error: %s", id, err)
-			return engerr.InternalServerError(err.Error())
-		}
+		var err error
+		ec, err = c.taskStartHelper(op, id, eid, name)
+		return err
+	}
+
+	if err := retry.DoWithConfig(operation, engerr.IsConflictError, backoffConf); err != nil {
+		op.Errorf("Failed to start Exec task for container(%s) due to error (%s)", id, err)
+		return err
+	}
+
+	attach := ec.OpenStdin || ec.OpenStdout || ec.OpenStderr
+
+	handle, err := c.Handle(id, name)
+	if err != nil {
+		op.Errorf("Failed to obtain handle during exec start for container(%s) due to error: %s", id, err)
+	}
+
+	// we need to be able to cancel it
+	taskOp, cancel := trace.WithCancel(&op, "exec task wait on %s", eid)
+	defer cancel()
+
+	attachResult := make(chan error, 1)
+	startResult := make(chan error, 1)
 
-		ec, err := c.containerProxy.InspectTask(op, handle, eid, id)
+	go func() {
+		// wait for the task to start
+		err := c.containerProxy.WaitTask(op, handle, name, eid)
```

Comment:
Should we use `taskOp` here instead of the parent `op`?

Reply:
yes we should be :) this change has been made. 👍